[CPDLP-1850] Refactor Accept and Reject NPQ application endpoints

leandroalemao opened the pull request:
### Context

Refactor `accept` and `reject` NPQ application endpoints in preparation for the new API v3

- Ticket: [CPDLP-1850](https://dfedigital.atlassian.net/browse/CPDLP-1850)

### Changes proposed in this pull request
Refactor API v1 & v2 `accept` and `reject` NPQ application endpoints in preparation for the new API v3 endpoint.

cwrw commented on app/controllers/api/v1/npq_applications_controller.rb:
```diff
@@ -26,19 +26,15 @@ def show
-        else
-          render json: { errors: Api::ErrorFactory.new(model: npq_application).call }, status: :bad_request
-        end
+        service = AcceptNPQApplication.new(npq_application:)
```
can we retain the `NPQ::AcceptApplication` structure, as this only relates to NPQ applications and no ECF? same for the reject, what do you think?

leandroalemao replied:
Hi @cwrw tks for your review. Addressed in https://github.com/DFE-Digital/early-careers-framework/pull/2674/commits/8c9fde545669f063b8a66cd2c524248a8025fdb5

cwrw commented on app/services/accept_npq_application.rb:
```diff
@@ -0,0 +1,105 @@
+  def not_already_accepted
+    return if npq_application.blank?
+
+    errors.add(:npq_application, I18n.t("activerecord.errors.models.npq_application.attributes.lead_provider_approval_status.has_already_been_accepted")) if npq_application.accepted?
```
could you move the error nesting in the locale to avoid the long path?

leandroalemao replied:
Hi @cwrw tks for your review. Addressed in https://github.com/DFE-Digital/early-careers-framework/pull/2674/commits/8c9fde545669f063b8a66cd2c524248a8025fdb5

cwrw commented on app/services/accept_npq_application.rb:
```diff
@@ -0,0 +1,105 @@
+      participant_profile
+      npq_application.update(lead_provider_approval_status: "accepted") && other_applications_in_same_cohort.update(lead_provider_approval_status: "rejected")
+      deduplicate_by_trn!
+      npq_application
```
can this move outside the transaction?

leandroalemao replied:
Hi @cwrw tks for your review. Addressed in https://github.com/DFE-Digital/early-careers-framework/pull/2674/commits/8c9fde545669f063b8a66cd2c524248a8025fdb5

cwrw commented on app/services/accept_npq_application.rb:
```diff
@@ -0,0 +1,105 @@
+      .where.not(id: npq_application.id)
+  end
+
+  def participant_profile
```
could this be renamed to indicate a profile is being created?

leandroalemao replied:
Hi @cwrw tks for your review. Addressed in https://github.com/DFE-Digital/early-careers-framework/pull/2674/commits/8c9fde545669f063b8a66cd2c524248a8025fdb5

cwrw commented on app/services/reject_npq_application.rb:
```diff
@@ -0,0 +1,34 @@
+  def not_already_rejected
+    return if npq_application.blank?
+
+    errors.add(:npq_application, I18n.t("activerecord.errors.models.npq_application.attributes.lead_provider_approval_status.has_already_been_rejected")) if npq_application.rejected?
```
same question here around error message, let me know your thoughts

cwrw commented on app/services/reject_npq_application.rb:
```diff
@@ -0,0 +1,34 @@
+  def not_already_rejected
+    return if npq_application.blank?
+
+    errors.add(:npq_application, I18n.t("activerecord.errors.models.npq_application.attributes.lead_provider_approval_status.has_already_been_rejected")) if npq_application.rejected?
```
same question here around error message, let me know your thoughts

leandroalemao replied:
Hi @cwrw tks for your review. Addressed in https://github.com/DFE-Digital/early-careers-framework/pull/2674/commits/8c9fde545669f063b8a66cd2c524248a8025fdb5

cwrw commented on spec/services/reject_npq_application_spec.rb:
```diff
@@ -0,0 +1,66 @@
+
+require "rails_helper"
+
+RSpec.shared_examples "validating an application to be rejected" do
```
do we need the shared examples as this is only for NPQ, same for other spec?

leandroalemao replied:
Hi @cwrw tks for your review. Addressed in https://github.com/DFE-Digital/early-careers-framework/pull/2674/commits/8c9fde545669f063b8a66cd2c524248a8025fdb5

cwrw reviewed (commented):
looking good, some minor comments inline. Also could you delete the old implementation please?

cwrw reviewed (approved):
Nicely done, thanks @leandroalemao

StupidCodeFactory commented on app/controllers/api/v1/npq_applications_controller.rb:
```diff
@@ -26,19 +26,15 @@ def show
-        else
-          render json: { errors: Api::ErrorFactory.new(model: npq_application).call }, status: :bad_request
-        end
+        service = NPQ::RejectApplication.new(npq_application:)
```
I would not have tied the action to the namespace but rather would have added classes as such:
```ruby
module NPQ
  module Application
    class Reject; end
    class Accept; end
  end
end
```

StupidCodeFactory commented on app/services/npq/accept_application.rb:
```diff
@@ -0,0 +1,112 @@
+        .joins(:participant_identity)
+        .where(participant_identity: { user_id: user.id })
+        .where(npq_course:)
+        .where(cohort:)
```
Could we collapse all these `#where` together (same for the method above):
```suggestion
        .where(cohort:, npq_course:, participant_identity: { user_id: user.id })
```

StupidCodeFactory commented on app/services/npq/accept_application.rb:
```diff
@@ -0,0 +1,112 @@
+    end
+
+    def teacher_profile
+      @teacher_profile ||= user.teacher_profile || user.build_teacher_profile
```
⚠️ ⚠️  Please change: Here you if the user does not have a teacher profile. 
You want to lookup teacher profile by by TRN and attach this `teacher_profile` to the user. In _theory_ there should not be two teacher profile with the same TRN. 
Another thing that needs to be though of is what if the user attach to the teacher profile is different than the one attached to the `NPQApplication` there should be some kind of error thrown as this has the potential to create duplicate profile!

@cwrw @peteryates @tonyheadford would you mind having a think about that, we've some some hacks in place to prevent double paying via the `TRN` routes but this looks like the place where duplicates could be unintentionally created?

At this point if it didn't exists you'd also want to add the TRN from `NPQApplication#teacher_reference_number` onto the `teacher_profile`

cwrw replied:
chatted with @StupidCodeFactory in person but for posterity: We'll check this in detail and create a ticket incase to address as it's out of scope for this PR (I'll link the ticket after investigation in the comments below)

StupidCodeFactory commented on app/services/npq/accept_application.rb:
```diff
@@ -0,0 +1,112 @@
+        .first
+        &.user
+
+      Identity::Transfer.call(from_user: participant_profile.user, to_user: same_trn_user) if same_trn_user
```
I've always been suspicious about this class and some potential unintended consequences. @cwrw could double check that this is ok. Lots of side effects. ⚠️

cwrw replied:
same here, will create a ticket after investigation to tackle issues raised in that class

StupidCodeFactory commented on app/services/npq/reject_application.rb:
```diff
@@ -0,0 +1,36 @@
+  class RejectApplication
+    include ActiveModel::Model
+    include ActiveModel::Attributes
+    include ActiveModel::Validations::Callbacks
```
✂️

StupidCodeFactory commented on app/services/npq/reject_application.rb:
```diff
@@ -0,0 +1,36 @@
+
+  private
+
+    def not_already_rejected
```
I think this validation is superfluous. If the Application is already rejected no need to return and error just skip the 
` npq_application.update!(lead_provider_approval_status: "rejected")`

StupidCodeFactory commented on spec/requests/api/v1/npq_applications_spec.rb:
```diff
@@ -267,20 +267,16 @@
-        expect(parsed_response["errors"][0].key?("detail")).to be_truthy
-        expect(parsed_response["errors"][0]["title"]).to eql("Status cannot change from accepted")
-        expect(parsed_response["errors"][0]["detail"]).to eql("Once accepted an application cannot change state")
+        expect(parsed_response.dig("errors", 0, "detail")).to eql("Once accepted an application cannot change state")
```
No biggy but we're losing coverage on the `title` attribute of the response.

leandroalemao commented:
Hi @StupidCodeFactory you ok? I've addressed some of the comments in https://github.com/DFE-Digital/early-careers-framework/pull/2674/commits/a3e05e5f8c1aa74a42d3d7e06bf6e86cc2340f4d tkssss
